vicodyn/src/vicodyn/darkmetrics.cpp

```diff
@@ -0,0 +1,172 @@
+                        continue;
+                    }
+                    peers_it->second.system_weight = system_usable_weight;
+                    COCAINE_LOG_DEBUG(logger_, "metrics for peer {}: system_usable_weight={} timestamp={}",
+                            peer.first, system_usable_weight, timestamp_s.count());
+                }
+            });
+
+            schedule_expiration();
+        } catch (const std::exception& err) {
+            COCAINE_LOG_ERROR(logger_, "failed to handle update for \"{}\" - {}", cluster_, err.what());
+            schedule_retry();
+        }
+    }
+
+    auto schedule_expiration() -> void {
+        expiration_timer_.expires_from_now(boost::posix_time::seconds(system_weights_ttl_.count()));
+        auto weak_parent = weak_parent_;
+        expiration_timer_.async_wait([&, weak_parent](const std::error_code& ec) {
+            const auto lock_parent = weak_parent.lock();
+            if (ec || !lock_parent) {
+                return;
+            }
+
+            COCAINE_LOG_WARNING(logger_, "timer expired, set system weight to {} for all peers in \"{}\"",
+                    defaults::system_weight, cluster_);
+            // Using of method apply_shared_unsafe is correct because peer_data_t::system_weight is atomic value
+            peers_.apply_shared_unsafe([&](peers_t::data_t& data) {
+                for (auto& peer_data : data.peers) {
+                    if (peer_data.second.peer->x_cocaine_cluster() != cluster_) {
+                        continue;
+                    }
+                    peer_data.second.system_weight = defaults::system_weight;
+                }
+            });
+        });
+    }
+
+    auto schedule_retry() -> void {
+        retry_timer_.expires_from_now(boost::posix_time::seconds(retry_after_.count()));
+        auto weak_parent = weak_parent_;
+        retry_timer_.async_wait([&, weak_parent](const std::error_code& ec) {
+            const auto lock_parent = weak_parent.lock();
+            if (ec || !lock_parent) {
+                return;
+            }
+
+            subscribe();
+            COCAINE_LOG_INFO(logger_, "resubscribed to metrics by path \"{}\"", path_);
+        });
+    }
+};
+
+darkmetrics_t::darkmetrics_t(context_t& context, peers_t& peers, const dynamic_t& args)
+    : unicorn_(api::unicorn(context, args.as_object().at("unicorn", "core").as_string()))
+    , unicorn_prefix_(args.as_object().at("unicorn_prefix", "/darkvoice").as_string())
+    , unicorn_retry_after_(args.as_object().at("unicorn_retry_after_s", 20U).as_uint())
+    , system_weights_enabled_(args.as_object().at("system_weights_enabled", true).as_bool())
+    , system_weights_ttl_(args.as_object().at("system_weights_ttl_s", 300U).as_uint())
+    , logger_(context.log("vicodyn/darkmetrics"))
+    , peers_(peers) {
+    COCAINE_LOG_INFO(logger_, "balancing by system weights is {}", system_weights_enabled_ ? "on" : "off");
+}
+
+darkmetrics_t::~darkmetrics_t() = default;
```

Comment:
Why you don't define it in header?

Reply:
It will not be compiling, because default destructor of `darkmetrics_t` require destructor of `std::list<subscribtion_t>`. But I used forward declaration for `subscribtion_t`.